// graphql-ts/src/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const requiredMetadataKey = Symbol("required");
const returnMetadataKey = Symbol("return");
const inputMetadataKey = Symbol("input");
const descriptionMetadataKey = Symbol("desc");
const nullableMetadatakey = Symbol("nullable");
function required(name) {
    return Reflect.metadata(requiredMetadataKey, JSON.stringify(name));
}
exports.required = required;
function returnType(name) {
    return Reflect.metadata(returnMetadataKey, JSON.stringify(name));
}
exports.returnType = returnType;
function description(text) {
    return Reflect.metadata(descriptionMetadataKey, text);
}
exports.description = description;
function inputListType(name) {
    return Reflect.metadata(inputMetadataKey, JSON.stringify(name));
}
exports.inputListType = inputListType;
function nullable(nullable) {
    return Reflect.metadata(nullableMetadatakey, nullable);
}
exports.nullable = nullable;
function getNullable(target, key) {
    var temp = Reflect.getMetadata(nullableMetadatakey, target, key);
    if (temp == undefined)
        return true;
    return temp;
}
exports.getNullable = getNullable;
function getInput(target, key) {
    var temp = Reflect.getMetadata(inputMetadataKey, target, key);
    if (!temp)
        return '';
    return temp.replace('"', '').replace('"', '');
}
exports.getInput = getInput;
function getDescription(target, key) {
    var temp = key ? Reflect.getMetadata(descriptionMetadataKey, target, key) : Reflect.getMetadata(descriptionMetadataKey, target);
    if (!temp)
        return '';
    return temp.replace('"', '').replace('"', '');
}
exports.getDescription = getDescription;
function getReturn(target, key) {
    var temp = Reflect.getMetadata(returnMetadataKey, target, key);
    if (!temp)
        return '';
    return temp.replace('"', '').replace('"', '');
}
exports.getReturn = getReturn;
function getRequired(target, key) {
    const temp = Reflect.getMetadata(requiredMetadataKey, target, key);
    if (temp) {
        return JSON.parse(temp);
    }
    else {
        return [];
    }
}
exports.getRequired = getRequired;
